spv/sync.go

```diff
@@ -540,6 +545,50 @@ func (s *Syncer) pickRemote(pick func(*p2p.RemotePeer) bool) (*p2p.RemotePeer, e
 	return nil, errors.E(errors.NoPeers)
 }
 
+// waitForAnyRemote blocks until there is one or more remote peers available or
+// the context is canceled.
+//
+// The [isEligible] callback determines which peers are eligible for selection.
+//
+// This function may return a nil peer with nil error if there are peers
+// available but none of them are eligible.
+func (s *Syncer) waitForRemote(ctx context.Context, isEligible func(rp *p2p.RemotePeer) bool) (*p2p.RemotePeer, error) {
+	for {
+		s.remotesMu.Lock()
+		if len(s.remotes) == 0 {
+			c := s.remoteAvailable
+			if c == nil {
+				c = make(chan struct{})
+				s.remoteAvailable = c
+			}
+			s.remotesMu.Unlock()
+
+			// Wait until a peer is available.
+			select {
+			case <-c:
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
+		} else {
+			eligible := make([]*p2p.RemotePeer, 0, len(s.remotes))
+			for _, rp := range s.remotes {
+				if isEligible(rp) {
+					eligible = append(eligible, rp)
+				}
+			}
+			s.remotesMu.Unlock()
+
+			if len(eligible) == 0 {
+				return nil, nil
+			}
+
+			// s.remotes is a map, therefore the first element in
+			// the slice is a random one.
```

Comment:
we wouldn't need to keep iterating the whole slice then, it could unlock the mutex and return the first eligible peer from the loop

Reply:
Right, I wrote it this way so that a future commit which introduced selecting the peer based on latency would be a smaller diff, but I think I see a way to make the "return on return on first eligible work" without building the secondary list. Will adjust.